File: 72. Edit Distance-2.cpp
class Solution {
public:
  int minDistance(string word1, string word2) {
    if (word1.size() < word2.size())
      swap(word1, word2);
    int m = word1.size(), n = word2.size();

    vector<int> dp(n + 1, 0);
    for (int j = 1; j <= n; ++j)
      dp[j] = j;

    for (int i = 1; i <= m; ++i) {
      int diag = dp[0];
      dp[0] = i;
      for (int j = 1; j <= n; ++j) {
        int up = dp[j];
        if (word1[i - 1] == word2[j - 1])
          dp[j] = diag;
        else {
          int I = dp[j - 1] + 1;
          int D = up + 1;
          int R = diag + 1;
          dp[j] = min(I, min(D, R));
        }
        diag = up;
      }
    }

    return dp[n];
  }
};

File: 72. Edit Distance-2.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)

        f = [[0] * (m + 1) for _ in range(n + 1)]
        f[0] = list(range(m + 1))

        for i, x in enumerate(word1):
            f[i + 1][0] = i + 1
            for j, y in enumerate(word2):
                if x == y:
                    f[i + 1][j + 1] = f[i][j]
                else:
                    f[i + 1][j + 1] = min(f[i + 1][j], min(f[i][j + 1], f[i][j])) + 1

        return f[n][m]
